docs/how-to-build-a-lexical-model.md

```diff
@@ -0,0 +1,365 @@
+be a single column of all of the words in the language.
+
+This is what my word list looks like in Google Sheets:
+
+![The word list, as it appears in Google Sheets](./sencoten-sheets-full.png).
+
+[sencoten-sheet]: https://docs.google.com/spreadsheets/d/10zhIc439BCSSooL_-HeJ6TUHd-ovkiXYcIGe-pHDTSg/edit?usp=sharing
+zsh:1: no matches found: https://docs.google.com/spreadsheets/d/10zhIc439BCSSooL_-HeJ6TUHd-ovkiXYcIGe-pHDTSg/edit?usp=sharing
+zsh:1: no matches found: https://docs.google.com/spreadsheets/d/10zhIc439BCSSooL_-HeJ6TUHd-ovkiXYcIGe-pHDTSg/edit?usp=sharing
+
+Now, we download the spreadsheet in the [required
+format](#appendix:tsv). To do this, in Google Sheets, select "File"
+» "Download as" » "Tab-separated values (.tsv, current sheet)".
+
+![Exporting the TSV file from Google Sheets](./sencoten-sheets-save-as.png)
+
+I'll save mine as **wordlist.tsv**.
+
+
+Step 2: Create a unique model ID
+--------------------------------
+
+Now that we have our TSV file, called **wordlist.tsv**, let's create the
+associated packaging folder and data.
+
+Our model needs a **unique ID**. The lexical model ID is a string in the
+format:
+
+> _author_._bcp47_._uniq_
+
+
+### Author
+
+_author_ is an identifier for the maintainer of the model. This
+identifier is a string of one or more ASCII lowercase letters, digits, and the
+underscore (`_`), and it must begin with a lowercase ASCII letter. Some
+organizations have a author identifier already, like `sil` for SIL
+International, `nrc` for the National Research Council Canada, and `fv`
+for FirstVoices.
+
+### BCP 47
+
+_bcp47_ is the [BCP 47][] tag of the language, script, and region you
+are modelling. If you need to uses dashes in your language's BCP 47 tag,
+replace them with underscores; likewise, replace uppercase letters with
+lowercase letters.
+
+Typically, this will just be the [ISO 639-3][] code of your language,
+and nothing else. In the case of SENĆOŦEN, [its ISO 639-3 code is
+`str`](https://iso639-3.sil.org/code/str).
+
+In some cases, you will have to specify the script (writing system), and
+the region. For example, if I'm creating a model for Plains Cree (ISO
+639-3: `crk`) it can be written both in the Latin script ([ISO 15924][]:
+`Latn`) and in Unified Canadian Aboriginal Syllabics ([ISO 15924][]:
+`Cans`). Say my model is only designed for syllabics. The correct BCP 47 tag
+is:
+
+    crk-Cans
+
+However, to create a valid lexical model ID, we need to replace the
+dash with an underscores, and convert all uppercase letters to
+lowercase. Thus, our `bcp47` is:
+
+    crk_cans
+
+[ISO 639-3]: https://iso639-3.sil.org/code_tables/639/data
+[ISO 15924]: https://en.wikipedia.org/wiki/ISO_15924
+[BCP 47]: https://tools.ietf.org/html/bcp47
+
+### Uniq
+
+_uniq_ is an arbitrary tag that uniquely identifies your specific model.
+This identifier can be anything, as long as it hasn't already been used
+before for this specific language. As with `author`
```

Comment:
Uniq only needs to be unique to the author, for that language; it doesn't technically need to be unique to other people.

Also note that this file naming scheme is only mandated for models in the shared repo. For private distribution you should be able to call it `foo` (even if that isn't recommended!)

Reply:
I'd find a more concrete example much more helpful. For example, Gene Rodgers of the National Research Council Canada is developing a lexical model for Straits Salish.
- author is `nrc` (for the National Research Council Canada)
- the BCP 47 code for the language is `str`
- the unique code is `grodgers` (or some other convention established by the NRC for uniquely identifying authors within the NRC)
- result is nrc.str.grdogers

Joe Bloggs is developing a lexical model for Hausa written with Arabic script.
- author is `bloggs`
- the BCP 47 code for the language is `ha_arab`
- the unique code is `joe` (or some other string that uniquely identifies Joe from other authors named `Bloggs`)
- result is bloggs.ha_arab.joe

Reply:
That's not quite right :)

The author is already unique -- if two developers named Bloggs turned up wanting to use the `bloggs` subtag, we'd ask the second one to pick a different author subtag, (so the second developer may get `alfredbloggs` for the author subtag. The `uniq` subtag is really about disambiguating multiple models for the same language from the same author, so for example:

Joe Bloggs has a Straits Salish model based on the Montler wordlist, and another one based on the Salish School Board wordlist (this is semi-fictional but ...). He wants both of them available because of different preferences in the community. So he picks a uniq subtag for each accordingly, for example:

* `bloggs.str.montler`
* `bloggs.str.schoolboard`

In practice, I hope there won't be many models for the same language. Note that dialect differences should really be picked up in the BCP 47 subtag (e.g. en_us, en_ca, en_au)!

Reply:
I appreciate the clarification. I'm still trying to wrap my head around this, and a concrete example helps -- especially when it shows that my understanding was wrong!!

Reply:
No problemo. The intent is to try to avoid running into namespacing issues which many projects (including keyman keyboards) have run into in the past. This may be a little over the top in some respects but isn't too hard, hopefully, for people to get their head around. The `uniq` subtag could also just be described as a name for the model.